src/Components/Components/src/ComponentBase.cs

```diff
@@ -177,188 +177,93 @@ void IComponent.Configure(RenderHandle renderHandle)
         }
 
-        private async Task ReRenderAsyncTask(Task task)
+        private async Task ProcessLifecycleTaskAsync(Task task)
         {
             try
             {
                 await task;
-                StateHasChanged();
             }
-            catch (Exception ex)
+            catch when (task.IsCanceled)
             {
-                // Either the task failed, or it was cancelled, or StateHasChanged threw.
-                // We want to report task failure or StateHasChanged exceptions only.
-                if (!task.IsCanceled)
-                {
-                    HandleException(ex);
-                }
-            }
-        }
-
-        private async void ContinueAfterLifecycleTask(Task task)
-        {
-            switch (task == null ? TaskStatus.RanToCompletion : task.Status)
-            {
-                // If it's already completed synchronously, no need to await and no
-                // need to issue a further render (we already rerender synchronously).
-                // Just need to make sure we propagate any errors.
-                case TaskStatus.RanToCompletion:
-                case TaskStatus.Canceled:
-                    break;
-                case TaskStatus.Faulted:
-                    HandleException(task.Exception);
-                    break;
-
-                // For incomplete tasks, automatically re-render on successful completion
-                default:
-                    try
-                    {
-                        await task;
-                        StateHasChanged();
-                    }
-                    catch (Exception ex)
-                    {
-                        // Either the task failed, or it was cancelled, or StateHasChanged threw.
-                        // We want to report task failure or StateHasChanged exceptions only.
-                        if (!task.IsCanceled)
-                        {
-                            HandleException(ex);
-                        }
-                    }
-
-                    break;
-            }
-        }
-
-        private static void HandleException(Exception ex)
-        {
-            if (ex is AggregateException && ex.InnerException != null)
-            {
-                ex = ex.InnerException; // It's more useful
+                // Ignore task cancelation but don't bother issuing a state change.
+                return;
             }
 
-            // TODO: Need better global exception handling
-            Console.Error.WriteLine($"[{ex.GetType().FullName}] {ex.Message}\n{ex.StackTrace}");
+            StateHasChanged();
         }
 
-        void IHandleEvent.HandleEvent(EventHandlerInvoker binding, UIEventArgs args)
+        async Task IHandleEvent.HandleEventAsync(EventHandlerInvoker binding, UIEventArgs args)
         {
-            var task = binding.Invoke(args);
-            ContinueAfterLifecycleTask(task);
+            await binding.Invoke(args);
```

Comment:
This is not correct. We want to trigger a render after calling binding.Invoke has finished executing the sync part and another optional render after the binding has completed executing in case it does actual async work.

This is similar to what we expect happen on SetParametersAsync.
A single render after all sync code has run
An additional render every time we switch between async/sync

Reply:
Ah yes!

@pranavkm Would be awesome if we could also have a unit test to represent the flow mentioned by @javiercn here.